package/MDAnalysis/analysis/base.py

```diff
@@ -125,14 +125,210 @@
 import inspect
 import logging
 import itertools
+import warnings
 
 import numpy as np
 from MDAnalysis import coordinates
 from MDAnalysis.core.groups import AtomGroup
 from MDAnalysis.lib.log import ProgressBar
+from MDAnalysis.lib.util import is_installed
+from typing import Callable, Iterable, Sequence
+from importlib import import_module
+from functools import partial
 
 logger = logging.getLogger(__name__)
 
+def is_valid_distributed_client(client: object):
+    try:
+        import dask.distributed as distributed
+    except ImportError:
+        return False
+    return isinstance(client, distributed.Client)
+
+class ParallelExecutor:
+    r"""Executor object that abstracts away running parallel computations
+    with various backends. Mainly is used for AnalysisBase, but can be used for other purposes as well."""
+
+    @classmethod
+    @property
+    def available_backends(cls):
```

Comment:
This is a crucial part of the new API. It really needs to be documented — I talked about it in https://github.com/MDAnalysis/mdanalysis/issues/4158#issuecomment-1694790525 but that's just what I assumed from reading the code.

I think my biggest conceptual problem is that **adding a new backend** (e.g., "mpi") will require **changing every single class in MDAnalysis and elsewhere**. This is impractical and defeats the idea of having a central AnalysisBase class that allows propagating of updates to user code. If I wrote an analysis class based on AnalysisBase then I should be able to profit from any changes.

I'll write this as a longer comment on the PR so that we can have more of a discussion.

Reply:
I guess this can also be resolved now